### PR TITLE
Add option to split example groups to individual examples for rspec

### PR DESCRIPTION
**Pull request by sbonebrake**
Add SPLIT_GROUPS env option for rspec tests to split example groups to run individual examples off of the queue.

### CODE REVIEW

**Comment by bootstraponline** [comment]
The other environment variables start with the `TEST_QUEUE_` prefix so this should probably be `TEST_QUEUE_SPLIT_GROUPS`

File: lib/test_queue/iterator.rb [hunk 1]
```diff
@@ -37,7 +37,8 @@ def each
           else
             yield suite
           end
-          @stats[suite.to_s] = Time.now - start
+          key = suite.respond_to?(:id) ? suite.id : suite
```
Comment (bootstraponline) [inline lib/test_queue/iterator.rb]: I think this is missing a `.to_s`?

`key = suite.respond_to?(:id) ? suite.id : suite.to_s`

File: lib/test_queue/runner/rspec.rb [hunk 2]
```diff
@@ -15,7 +15,15 @@ class Runner
     class RSpec < Runner
       def initialize
         @rspec = ::RSpec::Core::QueueRunner.new
-        super(@rspec.example_groups.sort_by{ |s| -(stats[s.to_s] || 0) })
+        @split_groups = ENV['SPLIT_GROUPS'] && ENV['SPLIT_GROUPS'].downcase == 'true'
```
Comment (bootstraponline) [inline lib/test_queue/runner/rspec.rb]: `ENV['SPLIT_GROUPS'].strip.downcase` might be better

**Comment by bootstraponline** [comment]
It'd be great to add a split groups [test](https://github.com/tmm1/test-queue/blob/master/test/sample_spec.rb) that verifies the expected speed improvement so that this feature doesn't regress in the future. I have a simple [benchmark example](https://github.com/bootstraponline/test_parallel) that may be useful.

**Comment by sbonebrake** [comment]
Fixed all of the issues except the test with a forced commit.

**Comment by bootstraponline** [comment]
Looks great! :+1:

**Comment by bootstraponline** [comment]
I have one more comment, the `environment variables` section of `readme.md` should be updated to include `TEST_QUEUE_SPLIT_GROUPS`.

**Comment by sbonebrake** [comment]
Ah, almost forgot the Readme. Updated that too.

File: lib/test_queue/runner/rspec.rb [hunk 3]
```diff
@@ -15,9 +15,24 @@ class Runner
     class RSpec < Runner
       def initialize
         @rspec = ::RSpec::Core::QueueRunner.new
-        super(@rspec.example_groups.sort_by{ |s| -(stats[s.to_s] || 0) })
+        @split_groups = ENV['TEST_QUEUE_SPLIT_GROUPS'] && ENV['TEST_QUEUE_SPLIT_GROUPS'].strip.downcase == 'true'
+        if @split_groups
+          groups = @rspec.example_groups
+          groups_to_split, groups_to_keep = [], []
+          groups.each do |group|
+            group.metadata[:no_split] ? groups_to_keep << group : groups_to_split << group
```
Comment (bootstraponline) [inline lib/test_queue/runner/rspec.rb]:
```ruby
(group.metadata[:no_split] ? groups_to_keep : groups_to_split) << group
```

**Comment by bootstraponline** [comment]
@tmm1 Let me know if any changes are required for this pull request to be merged.

**Comment by bootstraponline** [comment]
@tmm1 ping :smile:

**Comment by jenseng** [comment]
A couple critiques:

This can actually make the performance *worse* if you use `before(:all)` (say, for expensive db/fixture setup). It effectively turns it into a `before(:each)`, due to how `RSpec.world.filtered_examples` gets reset and the (filtered) example group gets (re-)run for each example.

A good middle ground might be to still just prioritize/assign at the group level, and enhance the group itself to pull stuff off the queue. Then you can ensure that specs within a given group run together and the `:all` hooks only run once per worker (while still being distributed across workers). Conceptually it's pretty simple, but I could see this getting kinda gnarly to hack into RSpec::Core::ExampleGroup.

I might take a stab at this, I'll post back with an update later with the results.

However this ends up, it would be nice to update lib/test_queue/runner/rspec2.rb as well, because right now this only works for RSpec 3

**Comment by jenseng** [comment]
Upon further investigation, while my proposed approach reduces the `before(:all)` cost, it's still not quite optimal... since every group could be distributed across every runner, they all could run each `before(:all)` once.

It seems like the ideal algorithm would still assign things out at the group level by default. Then if any runner runs out of stuff to do, it goes and helps finish up any big groups that are still being run. This has the potential to complicate the code quite a bit though :-/

**Comment by bootstraponline** [comment]
I think if you have a `before(:all)` then it's better to mark that describe with the `no_split` tag.

**Comment by bootstraponline** [comment]
> it would be nice to update lib/test_queue/runner/rspec2.rb as well, because right now this only works for RSpec 3

The APIs to make this work well on Sauce Labs require at least RSpec 3. RSpec 2 is legacy (last release June 1, 2014) so I'm not sure it's worth supporting. Is there a reason you can't update to RSpec 3?

**Comment by jenseng** [comment]
I'm using RSpec 3 in my projects, but test-queue supports RSpec 2. It seems like this pull request ought maintain that support unless there's a compelling reason not to ¯\_(ツ)_/¯

**Comment by sbonebrake** [comment]
Yes, for unit tests and tests that do a lot of work in a before :all block it is not a good idea to split them. However, we use them for 95% our selenium tests and add the :no_split tag to the other example groups.  I wouldn't say everyone should use this all the time, but it makes a huge difference for test suites like mine. I can look into adding it to RSpec 2.

**Comment by jenseng** [comment]
Yeah my idea is shaping up to be fairly involved to implement (requires hooks into ExampleGroup and some non-trivial changes to TestQueue::Iterator), I think it's outside the scope of this pull request... perhaps I'll stick it in an add-on gem.

As far as this PR goes, seems like it addresses the typical selenium use cases well :+1:

I think my only concern (besides RSpec 2) is that docs could be a little more clear for when and when they shouldn't be used, and the tradeoffs. Like maybe stick something like this before the bit about `:no_split`: "Note: because tests are divided up by example, context-level hooks like before(:all) will run with each example." ?

**Comment by tmm1** [comment]
Is this good to merge now?

**Comment by sbonebrake** [comment]
I've been using my fork with this commit for months now and it's been working fine. Let me know if there is anything else you want me to address

**Pull request merged**